src/drivers/imu/mpu9250/mpu9250.cpp

```diff
@@ -995,11 +1178,51 @@ MPU9250::gyro_ioctl(struct file *filp, int cmd, unsigned long arg)
 	}
 }
 
+int
+MPU9250::select_register_bank(uint8_t bank) {
+	uint8_t ret;
+	//if (bank != _selected_bank ) {
+	uint8_t buf;
+	uint8_t count=0;
+
+
+	ret = _interface->write(MPU9250_LOW_SPEED_OP(ICMREG_20948_BANK_SEL), &bank, 1);
+	if (ret != OK) {
+		return ret;
+	}
+	usleep(100);
+	_interface->read(MPU9250_LOW_SPEED_OP(ICMREG_20948_BANK_SEL), &buf, 1);
+
+	while(bank != buf) {
+		ret = _interface->write(MPU9250_LOW_SPEED_OP(ICMREG_20948_BANK_SEL), &bank, 1);
+		if (ret != OK) {
+			return ret;
+		}
+		_selected_bank=bank;
+
+		count++;
+		PX4_WARN("BANK retries: %d", count);
+
+
+		usleep(20);
```

Comment:
This doesn't work due to the scheduling quantum (minimum of one millisecond) and doesn't work in interrupt context. Furthermore 20 microseconds is quite low, just the switch from the write() to the read() call will take the CPU this time frame. Have you tried just leaving this out entirely?

Both udelay and usleep are problematic in operation. The way to do it is to re-schedule the driver to re-run, not to delay the whole system or thread.

Reply:
These are not very important as the success of switching the register bank is checked and re-tried. Tried to minimize checks and retries (with resulting I2C communication) by using delays here. I'll just take those out and re-think the other delays.These are not very important as the success of switching the register bank is checked and re-tried. Tried to minimize checks and retries (with resulting I2C communication) by using delays here. I'll just take those out and re-think the other delays.